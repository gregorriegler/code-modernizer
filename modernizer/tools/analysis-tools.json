[
  {
    "name": "entry-point-analysis",
    "description": "No description available",
    "arguments": [
      "project_path: string - Path to the project file",
      "file_name: string - Name of the file to analyze",
      "workspaceLoader: IWorkspaceLoader"
    ]
  }
]