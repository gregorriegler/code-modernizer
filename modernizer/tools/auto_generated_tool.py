from .base_auto_generated_tool import BaseAutoGeneratedTool

class AutoGeneratedTool(BaseAutoGeneratedTool):
    """
    A tool that automatically generates its interface from the corresponding C# refactoring tool
    using reflection to read the tool's arguments and description.
    """
    
    def __init__(self, tool_name, runcommand, tool_data):
        super().__init__(tool_name, runcommand, tool_data, "Refactoring")
        
    def _get_project_path(self):
        """Get the project path for refactoring tools"""
        return 'refactoring-tools/RoslynRefactoring/RoslynRefactoring.csproj'