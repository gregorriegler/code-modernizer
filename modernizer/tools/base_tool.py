import subprocess

class BaseTool:
    name = ''
    
    def __init__(self):
        self.runcommand = None
        
    def execute(self, *args, **kwargs):
        raise NotImplementedError("Subclasses must implement execute()")
    
    def get_description_from_csharp(self, tool_name):
        try:
            result = subprocess.run(
                ['dotnet', 'run', '--project', 'refactoring-tools/RoslynRefactoring/RoslynRefactoring.csproj', '--', '--describe', tool_name],
                capture_output=True,
                text=True,
                timeout=10
            )
            if result.returncode == 0:
                return result.stdout.strip()
            else:
                return f"Error getting description: {result.stderr}"
        except Exception as e:
            return f"Error getting description: {str(e)}"
