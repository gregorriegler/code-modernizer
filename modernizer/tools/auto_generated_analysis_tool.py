from .base_tool import BaseTool

class AutoGeneratedAnalysisTool(BaseTool):
    """
    A tool that automatically generates its interface from the corresponding C# analysis tool
    using reflection to read the tool's arguments and description.
    """
    
    def __init__(self, tool_name, runcommand, tool_data):
        super().__init__()
        self.name = tool_name
        self.runcommand = runcommand
        self._tool_data = tool_data
        self._info = self._convert_tool_data_to_info(tool_data)
        
    @property
    def info(self):
        """Get complete tool information from JSON data"""
        return self._info
    
    def _convert_tool_data_to_info(self, tool_data):
        """Convert JSON tool data to the expected info format"""
        args = tool_data.get('arguments', [])
        
        return {
            'name': tool_data.get('name', self.name),
            'description': tool_data.get('description', f'Analysis tool: {self.name}'),
            'arguments': args
        }
        
    @property
    def description(self):
        """Get the tool description from the C# analysis tool"""
        return self.info.get('description', f'Analysis tool: {self.name}')
    
    @property
    def arguments(self):
        """Get argument information"""
        info = self.info
        return info.get('arguments', [])
    
    def get_usage_info(self):
        """Get comprehensive usage information for this tool"""
        data = self._tool_data
        usage_info = f"""Tool: {data['name']}
Description: {data['description']}

Usage: /{data['name']} <arguments>

Arguments:"""
        
        for arg in data.get('arguments', []):
            usage_info += f"\n  - {arg}"
            
        return usage_info
    
    
    def execute(self, args):
        """Execute the analysis tool with the provided arguments"""
        arg_list = args.split() if isinstance(args, str) and args else []
        
        return self.runcommand('dotnet', [
            'run', '--project', 'refactoring-tools/RoslynAnalysis/RoslynAnalysis.csproj',
            '--', self.name
        ] + arg_list)