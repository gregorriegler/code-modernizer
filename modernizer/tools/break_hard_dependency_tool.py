from .base_tool import BaseTool

class BreakHardDependencyTool(BaseTool):
    name = 'break-hard-dependency'
    
    def __init__(self, runcommand):
        super().__init__()
        self.runcommand = runcommand
        
    def execute(self, args):
        arg_list = args.split()
        return self.runcommand('dotnet', ['run', '--project', 'refactoring-tools/RoslynRefactoring/RoslynRefactoring.csproj', '--', 'break-hard-dependency'] + arg_list)