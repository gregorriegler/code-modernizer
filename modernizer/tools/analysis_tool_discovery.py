import os
import subprocess
import json
from .auto_generated_analysis_tool import AutoGeneratedAnalysisTool

class AnalysisToolDiscovery:
    def __init__(self, runcommand):
        self.runcommand = runcommand
        self.cache_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'analysis-tools.json')
    
    def discover_analysis_tools(self):
        """Automatically discover all available C# analysis tools from JSON with caching"""
        tools_data = self._load_from_cache()
        if tools_data is None:
            tools_data = self._fetch_and_cache_tools()
        
        return [AutoGeneratedAnalysisTool(tool_data['name'], self.runcommand, tool_data)
                for tool_data in tools_data]
    
    def _load_from_cache(self):
        """Load tools data from cache file if it exists"""
        if os.path.exists(self.cache_file):
            with open(self.cache_file, 'r') as f:
                return json.load(f)
        return None
    
    def _fetch_and_cache_tools(self):
        """Fetch tools data from C# project and cache it"""
        tool_library_dir = os.path.dirname(os.path.abspath(__file__))
        project_path = os.path.join(tool_library_dir, '../../refactoring-tools/RoslynAnalysis/RoslynAnalysis.csproj')
        
        result = subprocess.run(
            ['dotnet', 'run', '--project', project_path, '--', '--list-tools'],
            capture_output=True,
            text=True,
            timeout=10
        )
        
        if result.returncode != 0:
            raise RuntimeError(f"Failed to discover analysis tools: {result.stderr}")
        
        tools_data = json.loads(result.stdout.strip())
        
        with open(self.cache_file, 'w') as f:
            json.dump(tools_data, f, indent=2)
        
        return tools_data