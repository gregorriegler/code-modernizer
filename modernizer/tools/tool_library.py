import re
import subprocess

from .ls_tool import LsTool
from .cat_tool import CatTool
from .test_tool import TestTool
from .coverage_tool import CoverageTool
from .mutation_tool import MutationTool
from .auto_generated_tool import AutoGeneratedTool
from .revert_tool import RevertTool
from .create_file_tool import CreateFileTool
from .edit_file_tool import EditFileTool
from .refactoring_tool_discovery import RefactoringToolDiscovery
from .analysis_tool_discovery import AnalysisToolDiscovery

class ToolLibrary:
    def __init__(self):
        static_tools = self._create_static_tools()
        dynamic_tools = self._discover_dynamic_tools()
        
        self.tools = static_tools + dynamic_tools
        self._build_tool_dict()
    
    def _create_static_tools(self):
        """Create the core static tools that are always available."""
        return [
            LsTool(self.runcommand),
            CatTool(self.runcommand),
            TestTool(self.runcommand),
            CoverageTool(self.runcommand),
            MutationTool(self.runcommand),
            RevertTool(self.runcommand),
            CreateFileTool(self.runcommand),
            EditFileTool(self.runcommand)
        ]
    
    def _discover_dynamic_tools(self):
        """Discover and create dynamic tools from external sources."""
        refactoring_discovery = RefactoringToolDiscovery(self.runcommand)
        refactoring_tools = refactoring_discovery.discover_refactoring_tools()
        
        analysis_discovery = AnalysisToolDiscovery(self.runcommand)
        analysis_tools = analysis_discovery.discover_analysis_tools()
        
        return refactoring_tools + analysis_tools
    
    def _build_tool_dict(self):
        self.tool_dict = {tool.name: tool for tool in self.tools}
    
    def get_tool_info(self, tool_name=None):
        """Get comprehensive information about tools"""
        if tool_name:
            tool = self.tool_dict.get(tool_name)
            if not tool:
                return f"Tool '{tool_name}' not found. Available tools: {', '.join(self.tool_dict.keys())}"
            
            if hasattr(tool, 'get_usage_info'):
                return tool.get_usage_info()
            else:
                return f"Tool: {tool.name}\nDescription: {getattr(tool, 'description', 'No description available')}"
        else:
            info_lines = ["Available Tools:"]
            for tool in self.tools:
                description = getattr(tool, 'description', 'No description available')
                info_lines.append(f"  {tool.name}: {description}")
            return "\n".join(info_lines)
    
    def get_refactoring_tools_info(self):
        """Get detailed information about all refactoring tools"""
        refactoring_tools = [tool for tool in self.tools if isinstance(tool, AutoGeneratedTool)]
        if not refactoring_tools:
            return "No refactoring tools found."
        
        info_lines = ["Refactoring Tools (auto-discovered from C#):"]
        for tool in refactoring_tools:
            info_lines.append(f"\n{tool.get_usage_info()}")
        
        return "\n".join(info_lines)
        
    def parse_and_execute(self, text):
        pattern = r'^/([\w-]+)(?:\s+(.*))?$'
        
        # Handle multi-line commands by treating the entire text as one command
        if '\n' in text:
            # For multi-line text, extract command from first line but use entire text for arguments
            first_line = text.splitlines()[0].strip()
            match = re.match(pattern, first_line)
            if match:
                cmd, _ = match.groups()
                tool = self.tool_dict.get(cmd)
                if tool:
                    # Extract arguments from the entire text, preserving newlines
                    # Find the position after "/{cmd} " in the text
                    cmd_prefix = f"/{cmd} "
                    if text.startswith(cmd_prefix):
                        full_args = text[len(cmd_prefix):]
                        result = tool.execute(full_args.strip() if full_args else None)
                        return text, result['output']
                    else:
                        result = tool.execute(None)
                        return text, result['output']
                else:
                    return text, f"Unknown command: {cmd}"
        
        # Single-line processing (original behavior)
        for line in text.splitlines():
            line = line.strip()
            match = re.match(pattern, line)
            if match:
                cmd, arg = match.groups()
                tool = self.tool_dict.get(cmd)
                if tool:
                    result = tool.execute(arg.strip() if arg else None)
                    return line, result['output']
                else:
                    return line, f"Unknown command: {cmd}"
        return text.splitlines()[0].strip(), ""

    def runcommand(self, cmd, args=None, cwd=None):
        try:
            if args:
                if isinstance(args, str):
                    args = [args]
                command_line = [cmd] + args
    
            timeout = 300 if (args and 'stryker' in args) else 30
            
            result = subprocess.run(
                command_line,
                capture_output=True,
                text=True,
                encoding='utf-8',
                errors='replace',
                timeout=timeout,
                cwd=cwd
            )
    
            output = result.stdout.rstrip('\n')
            
            if result.stderr:
                stderr = result.stderr
                if output:  # Only add newline if there's stdout content
                    output += f"\nSTDERR: {stderr}"
                else:
                    output = f"STDERR: {stderr}"
    
            return {
                'success': result.returncode == 0,
                'output': output,
                'returncode': result.returncode
            }
        except subprocess.TimeoutExpired:
            return {'success': False, 'output': 'Command timed out (30s limit)', 'returncode': -1}
        except Exception as e:
            return {'success': False, 'output': f'Error: {str(e)}', 'returncode': -1}
