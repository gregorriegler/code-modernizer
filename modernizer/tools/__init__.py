from .base_tool import BaseTool
from .tool_library import ToolLibrary
from .cat_tool import CatTool
from .ls_tool import LsTool
from .revert_tool import RevertTool
from .test_tool import TestTool

__all__ = [
    'BaseTool',
    'ToolLibrary',
    'CatTool',
    'LsTool',
    'RevertTool',
    'TestTool'
]