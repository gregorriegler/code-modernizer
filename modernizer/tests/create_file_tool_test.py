import os
from contextlib import contextmanager
from approvaltests import Options, verify
from approvaltests.scrubbers import combine_scrubbers

from modernizer.tools.tool_library import ToolLibrary
from .test_helpers import create_path_scrubber, create_date_scrubber, create_ls_error_scrubber

library = ToolLibrary()

@contextmanager
def temp_directory(tmp_path):
    """Context manager to handle directory changes for tests."""
    original_cwd = os.getcwd()
    os.chdir(tmp_path)
    try:
        yield
    finally:
        os.chdir(original_cwd)

def verifyCreateTool(framework, command, expected_filename, expected_content=None, tmp_path=None):
    """
    Verify a create tool command and check the created file's content.
    
    Args:
        framework: The tool framework to execute the command
        command: The command to execute
        expected_filename: The filename that should be created
        expected_content: The expected content of the file (None for empty file)
        tmp_path: Optional temporary path to change to during execution
    """
    if expected_content is None:
        expected_content = ""
    
    # Create scrubber for consistent output
    path_scrubber = create_path_scrubber()
    date_scrubber = create_date_scrubber()
    ls_error_scrubber = create_ls_error_scrubber()
    multi_scrubber = combine_scrubbers(path_scrubber, date_scrubber, ls_error_scrubber)
    
    def _execute_and_verify():
        cmd, result = framework.parse_and_execute(command)
        
        # Check if file was created
        if not os.path.exists(expected_filename):
            raise AssertionError(f"File '{expected_filename}' should have been created")
        
        # Read and verify file content
        with open(expected_filename, "r") as f:
            actual_content = f.read()
        
        if actual_content != expected_content:
            raise AssertionError(f"File '{expected_filename}' should contain '{expected_content}', but contains '{actual_content}'")
        
        # Create verification output including file content
        file_info = f"File created: {expected_filename}\nFile content:\n--- FILE CONTENT START ---\n{actual_content}\n--- FILE CONTENT END ---"
        verify(f"Command:\n{cmd}\n\nResult:\n{result}\n\n{file_info}", options=Options().with_scrubber(multi_scrubber))
    
    if tmp_path:
        with temp_directory(tmp_path):
            _execute_and_verify()
    else:
        _execute_and_verify()

def test_create_tool_single_character_name(tmp_path):
    verifyCreateTool(library, "/create-file a", "a", tmp_path=tmp_path)

def test_create_tool_simple_name_with_extension(tmp_path):
    verifyCreateTool(library, "/create-file test.txt", "test.txt", tmp_path=tmp_path)

def test_create_tool_single_character_content(tmp_path):
    verifyCreateTool(library, "/create-file test.txt a", "test.txt", "a", tmp_path=tmp_path)

def test_create_tool_simple_text_content(tmp_path):
    verifyCreateTool(library, "/create-file readme.txt Hello World", "readme.txt", "Hello World", tmp_path=tmp_path)

def test_create_tool_newline_content(tmp_path):
    verifyCreateTool(library, '/create-file multi.txt "Line 1\\nLine 2"', "multi.txt", "Line 1\nLine 2", tmp_path=tmp_path)

def test_create_tool_json_content(tmp_path):
    verifyCreateTool(library, '/create-file config.json {"name": "test"}', "config.json", '{"name": "test"}', tmp_path=tmp_path)

def test_create_tool_explicit_empty_content(tmp_path):
    verifyCreateTool(library, '/create-file empty.txt ""', "empty.txt", "", tmp_path=tmp_path)

def test_create_file_in_nonexistent_directory(tmp_path):
    verifyCreateTool(library, f'/create-file src/utils/helper.py "# Helper module"', "src/utils/helper.py", "# Helper module", tmp_path=tmp_path)

def test_create_file_permission_error():
    """Test creating file in system directory without permissions should return error message"""
    # Use a path that should definitely cause permission error on Windows
    cmd, result = library.parse_and_execute("/create-file C:/Windows/System32/test.txt")
    
    # Should fail with permission error - result is a string containing the error
    assert 'permission' in result.lower() or 'denied' in result.lower() or 'error' in result.lower()

def test_create_file_already_exists(tmp_path):
    """Test creating file that already exists should return error message"""
    with temp_directory(tmp_path):
        # First create a file
        library.parse_and_execute("/create-file existing.txt")
        
        # Try to create the same file again
        cmd, result = library.parse_and_execute("/create-file existing.txt")
        
        # Should fail with "file already exists" error
        assert 'already exists' in result.lower() or 'exists' in result.lower()

def test_create_file_invalid_filename():
    """Test creating file with invalid characters should return error message"""
    cmd, result = library.parse_and_execute('/create-file "file<>name.txt"')
    
    # Should fail with invalid filename error
    assert 'invalid' in result.lower() or 'error' in result.lower()