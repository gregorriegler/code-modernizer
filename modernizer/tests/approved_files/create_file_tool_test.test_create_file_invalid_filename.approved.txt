Command:
/create-file "file<>name.txt"

Result:
Error creating file '"file<>name.txt"': [Errno 22] Invalid argument: '"file<>name.txt"'

No File created
